cc/mcts_node.cc

```diff
@@ -0,0 +1,254 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "cc/mcts_node.h"
+
+#include <algorithm>
+#include <cmath>
+#include <functional>
+#include <iomanip>
+#include <sstream>
+#include <tuple>
+#include <utility>
+
+#include "cc/algorithm.h"
+
+namespace minigo {
+
+constexpr float MctsNode::kPuct;
+
+MctsNode::MctsNode(EdgeStats* stats, const Position& position)
+    : parent(nullptr), stats(stats), move(Coord::kInvalid), position(position) {
+  for (int i = 0; i < kNumMoves; ++i) {
+    illegal_moves[i] = position.IsMoveLegal(i) ? 0 : 1000;
+  }
+  DualNet::InitializeFeatures(position, &features);
+}
+
+MctsNode::MctsNode(MctsNode* parent, Coord move)
+    : parent(parent),
+      stats(&parent->edges[move]),
+      move(move),
+      position(parent->position) {
+  position.PlayMove(move);
+  for (int i = 0; i < kNumMoves; ++i) {
+    illegal_moves[i] = position.IsMoveLegal(i) ? 0 : 1000;
+  }
+  DualNet::UpdateFeatures(parent->features, position, &features);
+}
+
+std::string MctsNode::Describe() const {
+  using std::setprecision;
+  using std::setw;
+
+  auto child_action_score = CalculateChildActionScore();
+  using SortInfo = std::tuple<float, float, int>;
+  std::array<SortInfo, kNumMoves> sort_order;
+  for (int i = 0; i < kNumMoves; ++i) {
+    sort_order[i] = {child_N(i), child_action_score[i], i};
+  }
+  std::sort(sort_order.begin(), sort_order.end(), std::greater<SortInfo>());
+
+  std::ostringstream oss;
+  oss << std::fixed;
+  oss << setprecision(4) << Q() << "\n";
+  oss << MostVisitedPath() << "\n";
+  oss << "move : action     Q      U      P    P-Dir    N  soft-N  p-delta  "
+         "p-rel\n";
+
+  float child_N_sum = 0;
+  for (const auto& e : edges) {
+    child_N_sum += e.N;
+  }
+  for (int rank = 0; rank < 15; ++rank) {
+    Coord i = std::get<2>(sort_order[rank]);
+    if (child_N(i) == 0) {
+      break;
+    }
+    float soft_N = child_N(i) / child_N_sum;
+    float p_delta = soft_N - child_P(i);
+    float p_rel = p_delta / child_P(i);
+    // clang-format off
+    oss << std::left << setw(5) << i.ToKgs() << std::right
+        << ": " << setw(6) << setprecision(3) << child_action_score[i]
+        << ", " << setw(6) << child_Q(i)
```

Comment:
you don't actually need to propagate my inconsistent string formatting into this version.  sorry :(